src/common/utxobased/db/makeProcessor.ts

```diff
@@ -193,25 +193,49 @@ export async function makeProcessor(
 
     async saveUtxo(utxo: IUTXO): Promise<void> {
       return await baselets.utxo(async tables => {
-        await tables.utxoIdsByScriptPubkey.insert(
-          '',
-          utxo.scriptPubkey,
-          utxo.id
-        )
+        const [utxoIds = []] = await tables.utxoIdsByScriptPubkey.query('', [
+          utxo.scriptPubkey
+        ])
+        if (!utxoIds.includes(utxo.id)) {
+          utxoIds.push(utxo.id)
+          await tables.utxoIdsByScriptPubkey.insert(
+            '',
+            utxo.scriptPubkey,
+            utxoIds
+          )
+        }
 
         await tables.utxoById.insert('', utxo.id, utxo)
       })
     },
 
     async removeUtxos(utxoIds: string[]): Promise<void> {
       return await baselets.utxo(async tables => {
+        // To remove utxo from utxoIdsByScriptPubkey table:
+        // 1. Create a map of scriptPubkeys to utxoIds
+        // 2. Use the map to remove utxoIds from the utxoIdsByScriptPubkey table
         const utxos = (await tables.utxoById.query('', utxoIds)).filter(
           utxo => utxo != null
         ) as IUTXO[]
-        await tables.utxoIdsByScriptPubkey.delete(
-          '',
-          utxos.map(utxo => utxo.scriptPubkey)
-        )
+        const utxoIdsMap: { [scriptPubkey: string]: string[] } = {}
+        for (const utxo of utxos) {
+          utxoIdsMap[utxo.scriptPubkey] = [
+            ...(utxoIdsMap[utxo.scriptPubkey] ?? []),
+            utxo.id
+          ]
+        }
+        for (const scriptPubkey of Object.keys(utxoIdsMap)) {
+          const utxoIdsToRemove = utxoIdsMap[scriptPubkey]
+          const [utxoIds = []] = await tables.utxoIdsByScriptPubkey.query('', [
+            scriptPubkey
+          ])
+          for (const utxoIdToRemove of utxoIdsToRemove) {
+            utxoIds.splice(utxoIds.indexOf(utxoIdToRemove), 1)
+          }
+          await tables.utxoIdsByScriptPubkey.insert('', scriptPubkey, utxoIds)
```

Comment:
Is this missing the case where no utxoIds remain? Shouldn't this be something like:
```
if (Array.length(utxoIds) > 0)
    await tables.utxoIdsByScriptPubkey.insert('', scriptPubkey, utxoIds)
else 
    await tables.utxoIdsByScriptPubkey.delete(''', scriptPubkey)
```

Reply:
I'm assuming you mean to say that we should delete the database entry once the `utxoIds` in the database becoomes empty. This is only a bit of extra cleanup on the hard-disk. After giving it some thought, I'm for this improvement.